scipy-stubs/signal/_peak_finding.pyi

```diff
@@ -1,37 +1,72 @@
-def argrelextrema(data: Untyped, comparator: Untyped, axis: int = 0, order: int = 1, mode: str = "clip") -> Untyped: ...
-def peak_prominences(x: Untyped, peaks: Untyped, wlen: Untyped | None = None) -> Untyped: ...
+_Mode: TypeAlias = Literal["clip", "wrap"]
+_ProminencesResult: TypeAlias = tuple[npt.NDArray[np.float64], npt.NDArray[np.intp], npt.NDArray[np.intp]]
+_WidthsResult: TypeAlias = tuple[
+    npt.NDArray[np.float64], npt.NDArray[np.float64], npt.NDArray[np.float64], npt.NDArray[np.float64]
+]
+
+class _FindPeaksResultsDict(TypedDict):
+    peak_heights: NotRequired[npt.NDArray[np.float64]]
+    left_thresholds: NotRequired[npt.NDArray[np.float64]]
+    right_thresholds: NotRequired[npt.NDArray[np.float64]]
+    prominences: NotRequired[npt.NDArray[np.float64]]
+    left_bases: NotRequired[npt.NDArray[np.intp]]
+    right_bases: NotRequired[npt.NDArray[np.intp]]
+    width_heights: NotRequired[npt.NDArray[np.float64]]
+    left_ips: NotRequired[npt.NDArray[np.float64]]
+    right_ips: NotRequired[npt.NDArray[np.float64]]
+    plateau_sizes: NotRequired[npt.NDArray[np.intp]]
+    left_edges: NotRequired[npt.NDArray[np.intp]]
+    right_edges: NotRequired[npt.NDArray[np.intp]]
+
+def argrelmin(
+    data: npt.NDArray[np.generic], axis: int = 0, order: int = 1, mode: _Mode = "clip"
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def argrelmax(
+    data: npt.NDArray[np.generic], axis: int = 0, order: int = 1, mode: _Mode = "clip"
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def argrelextrema(
+    data: npt.NDArray[np.generic],
+    comparator: Callable[[npt.NDArray[np.generic], npt.NDArray[np.generic]], npt.NDArray[np.bool_]],
+    axis: int = 0,
+    order: int = 1,
+    mode: _Mode = "clip",
+) -> tuple[npt.NDArray[np.intp], ...]: ...
+def peak_prominences(x: npt.ArrayLike, peaks: npt.ArrayLike, wlen: int | None = None) -> _ProminencesResult: ...
 def peak_widths(
-    x: Untyped,
-    peaks: Untyped,
+    x: npt.ArrayLike,
+    peaks: npt.ArrayLike,
     rel_height: float = 0.5,
-    prominence_data: Untyped | None = None,
-    wlen: Untyped | None = None,
-) -> Untyped: ...
+    prominence_data: _ProminencesResult | None = None,
+    wlen: int | None = None,
+) -> _WidthsResult: ...
 def find_peaks(
-    x: Untyped,
-    height: Untyped | None = None,
-    threshold: Untyped | None = None,
-    distance: Untyped | None = None,
-    prominence: Untyped | None = None,
-    width: Untyped | None = None,
-    wlen: Untyped | None = None,
+    x: npt.ArrayLike,
+    height: float | npt.NDArray[np.float64] | tuple[float | None, float | None] | None = None,
+    threshold: float | npt.NDArray[np.float64] | tuple[float | None, float | None] | None = None,
+    distance: np.float64 | None = None,
+    prominence: float | npt.NDArray[np.float64] | tuple[float | None, float | None] | None = None,
+    width: float | npt.NDArray[np.float64] | tuple[float | None, float | None] | None = None,
```

Comment:
I'm guessing that when `float` is allowed, things like `np.float16` and `np.bool_` are also allowed. 
And in cases like these were I'm feeling lazy and don't want to verify this, I usually play it safe and just use `scipy._typing.AnyReal` for it. But in my experience, it's very rare that *only* `float` is alowed, so this usually holds.

Reply:
Similarly, I also suspect that `npt.NDArray[np.float64]` is also a bit too narrow, and that all other scalar types that can be coerced to `float64` are also allowed, so e.g. `float16`, `uint8` and `bool_` are also allowed (if my guess is right in this case)

Reply:
> I'm guessing that when `float` is allowed, things like `np.float16` and `np.bool_` are also allowed. And in cases like these were I'm feeling lazy and don't want to verify this, I usually play it safe and just use `scipy._typing.AnyReal` for it. But in my experience, it's very rare that _only_ `float` is alowed, so this usually holds.

Thanks will change that. In terms of the source code I think usually these arguments must support comparison with certain arrays via the function `_select_by_property` after it is unpacked which is an operation which preserves their type/dtype.

```python
def _select_by_property(peak_properties, pmin, pmax):
    keep = np.ones(peak_properties.size, dtype=bool)
    if pmin is not None:
        keep &= (pmin <= peak_properties)
    if pmax is not None:
        keep &= (peak_properties <= pmax)
    return keep
```

This might be a bit too narrow/specific of a typing though in case maintainers want to do something more with the types later on.